pkg/controller/snapshotter/snapshotter.go

```diff
@@ -160,94 +165,87 @@ func (vs *volumeSnapshotter) getSnapshotDataFromSnapshotName(snapshotName string
-			break
-		}
-	}
-
-	if completed {
-		glog.Infof("Snapshot %s successfully created and it is ready.", snapshotName)
-		// UpdateVolmeSnapshot
-		snapConditions := []crdv1.VolumeSnapshotCondition{
-			{
-				Type:               crdv1.VolumeSnapshotConditionReady,
-				Status:             v1.ConditionTrue,
-				Message:            "Snapshot created succsessfully",
-				LastTransitionTime: metav1.Now(),
-			},
-		}
-		_, err = vs.UpdateVolumeSnapshot(snapshotName, &snapConditions, true)
-		if err != nil {
-			glog.Errorf("Error updating volume snapshot %s: %v", snapshotName, err)
-			// NOTE(xyang): Return error if failed to update VolumeSnapshot after
-			// create snapshot request is sent to the plugin and VolumeSnapshotData is updated
-			return fmt.Errorf("Failed to update VolumeSnapshot for snapshot %s: %v", snapshotName, err)
-		}
-
-		snapDataConditions := []crdv1.VolumeSnapshotDataCondition{
-			{
-				Type:               crdv1.VolumeSnapshotDataConditionReady,
-				Status:             v1.ConditionTrue,
-				Message:            "Snapshot data created succsessfully",
-				LastTransitionTime: metav1.Now(),
-			},
+			continue
 		}
-
-		// Update VolumeSnapshotData status
-		err = vs.UpdateVolumeSnapshotData(snapshotDataName, &snapDataConditions)
-		if err != nil {
+		glog.Infof("waitForPluginSnapshot: Snapshot %s creation is complete: %#v", snapshotName, conditions)
+
+                _, err = vs.UpdateVolumeSnapshot(snapshotName, conditions)
+                if err != nil {
+                        glog.Errorf("Error updating volume snapshot %s: %v", snapshotName, err)
+                        // NOTE(xyang): Return error if failed to update VolumeSnapshot after
+                        // create snapshot request is sent to the plugin and VolumeSnapshotData is updated
+                        return fmt.Errorf("Failed to update VolumeSnapshot for snapshot %s: %v", snapshotName, err)
+                }
+
+                ind := len(*conditions) - 1
+                snapDataConditions := []crdv1.VolumeSnapshotDataCondition{
+                        {
+                                Type:               (crdv1.VolumeSnapshotDataConditionType)((*conditions)[ind].Type),
+                                Status:             (*conditions)[ind].Status,
+                                Message:            (*conditions)[ind].Message,
+                                LastTransitionTime: metav1.Now(),
+                        },
+                }
+
+                // Update VolumeSnapshotData status
+                err = vs.UpdateVolumeSnapshotData(snapshotDataName, &snapDataConditions)
+                if err != nil {
 			return fmt.Errorf("Error update snapshotData object %s: %v", snapshotName, err)
-		}
+                }
 
-		glog.Infof("waitForPluginSnapshot: Snapshot %s created successfully. Adding it to Actual State of World.", snapshotName)
-		vs.actualStateOfWorld.AddSnapshot(snapshot)
-		return nil
+		if newstatus == statusReady {
```

Comment:
if statusError, we should also return but with error instead of going on to the next loop?

Reply:
Sure

Reply:
Addressed on line 242-243.